src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -97,37 +153,86 @@ private bool ApplicableOperatorIs(MethodSymbol candidate, CSharpSyntaxNode node,
         {
             bool wasExpression;
-            return BindConstantPattern(node, operand, operandType, node.Expression, hasErrors, diagnostics, out wasExpression, wasSwitchCase);
+            return BindConstantPattern(node, operand, operandIsNull, operandType, node.Expression, hasErrors, diagnostics, out wasExpression, wasSwitchCase);
         }
 
-        internal BoundPattern BindConstantPattern(
+        internal BoundConstantPattern BindConstantPattern(
             CSharpSyntaxNode node,
-            BoundExpression left,
-            TypeSymbol leftType,
-            ExpressionSyntax right,
+            BoundExpression operand,
+            bool? operandIsNull,
+            TypeSymbol operandType,
+            ExpressionSyntax patternExpression,
             bool hasErrors,
             DiagnosticBag diagnostics,
             out bool wasExpression,
             bool wasSwitchCase)
         {
-            var expression = BindValue(right, diagnostics, BindValueKind.RValue);
+            var expression = BindValue(patternExpression, diagnostics, BindValueKind.RValue);
+            ConstantValue constantValueOpt = null;
+            expression = ConvertPatternExpression(operandType, patternExpression, expression, ref constantValueOpt, diagnostics);
             wasExpression = expression.Type?.IsErrorType() != true;
-            if (!node.HasErrors && expression.ConstantValue == null)
+            if (!expression.HasErrors && constantValueOpt == null)
             {
-                diagnostics.Add(ErrorCode.ERR_ConstantExpected, right.Location);
+                diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location);
                 hasErrors = true;
             }
 
-            return new BoundConstantPattern(node, expression, hasErrors);
+            bool? knownMatchResult = null;
+            if (constantValueOpt != null)
+            {
+                if (constantValueOpt.IsNull)
+                {
+                    knownMatchResult = operandIsNull;
+                }
+                else if (operand.ConstantValue != null)
+                {
+                    knownMatchResult = constantValueOpt.Equals(operand.ConstantValue);
+                }
+            }
+
+            return new BoundConstantPattern(node, expression, constantValueOpt, knownMatchResult, hasErrors);
+        }
+
+        internal BoundExpression ConvertPatternExpression(TypeSymbol leftType, CSharpSyntaxNode node, BoundExpression expression, ref ConstantValue constantValue, DiagnosticBag diagnostics)
+        {
+            // NOTE: This will allow user-defined conversions, even though they're not allowed here.  This is acceptable
+            // because the result of a user-defined conversion does not have a ConstantValue and we'll report a diagnostic
+            // to that effect later.
+            BoundExpression convertedCaseExpression = GenerateConversionForAssignment(leftType, expression, diagnostics);
```

Comment:
Is this function meant to be called only for case expressions? If so, please change the name of the function accordingly. Otherwise, please remove ```Case``` from the name of this local in order to avoid confusing the reader.